ouroboros-consensus-test/src/Test/Ouroboros/Consensus/ChainGenerator/Honest.hs

```diff
@@ -0,0 +1,638 @@
+
+        start <- BV.testMV S.notInverted mv (C.windowStart ehcgSlots)
+        writeSTRef rtot $! (if start then subtract 1 else id) $ tot'
+
+    pure mv
+  where
+    UnsafeCheckedHonestRecipe {
+        chrEhcgDensity
+      ,
+        chrWin         = slots
+      }  = recipe
+
+    sz  = C.windowSize slots :: C.Size hon SlotE   -- ie 'Len'
+
+    wrap v = ChainSchedule slots v
+
+-----
+
+-- | The population count of a sample drawn from the stationary distribution of
+-- the idealized honest chain generator
+genFirstWindowPopCount :: R.RandomGen g => Asc -> BV.SomeDensityWindow S.NotInverted -> g -> (Int, g)
+genFirstWindowPopCount asc densityWindow g =
+    let q :: Double
+        (q, g') = R.random g   -- note that q is between 0 and 1.
+
+        cumul = tail $ scanl (+) 0 $ map abs $ Mat.toList sd
+
+        -- A sample drawn from the weighted categorical distribution that is
+        -- the Markov chain's stationary distribution.
+        i = length $ takeWhile (\x -> x / last cumul < q) cumul
+    in
+    (n + i, g')
+  where
+    f = ascVal asc
+
+    BV.SomeDensityWindow (C.Count n) (C.Count d) = densityWindow
+
+    sz = d - n + 1
+
+    dd = toEnum d :: Double
+    nn = toEnum n :: Double
+
+    -- the non-normalized stationary distribution
+    --
+    -- 1) See
+    -- https://stephens999.github.io/fiveMinuteStats/markov_chains_discrete_stationary_dist.html
+    --
+    -- 2) The stationary distribution is the eigenvector with eigenvalue 1.
+    --
+    -- 3) We want the /left/ eigenvector (TODO is this fundamental to transition matrices?).
+    --
+    -- Thus we want a vector v such that
+    --
+    -- > v A       = v
+    -- > v A  - v  = 0
+    -- > v (A - I) = 0
+    --
+    -- We find that using 'Mat.null1' of the transpose of A - I.
+    sd :: Mat.Vector Double
+    sd = Mat.null1 $ Mat.tr $ p - Mat.ident sz
+
+{-
+    This would be equally (?) exact, but it makes the test suite about 2-3
+    times slower than does 'Mat.null1'.
+
+    sd :: Mat.Vector Double
+    sd = lvecs Mat.! Mat.maxIndex vals
+
+    (Mat.cmap realPart -> vals, Mat.tr -> Mat.cmap realPart -> lvecs) = Mat.eig (Mat.tr p)
+-}
+
+{-
+    TODO This makes the test suite 3-5 times faster than does 'Mat.null1', but
+    I don't know how to bound its inaccuracy. Maybe there's an argument that
+    for some X after X * (s-d) steps (instead of 1024) it's certainly "close
+    enough" to the stationary distribution for our specific use case?
+
+    -- a brute-force way to (hopefully) find the stationary distribution
+    -- (recall the way in which 'stimes' is optimized) sd :: Mat.Vector Double
+    -- sd = (Mat.! 0) $ stimes (2^(10 :: Int) :: Int) p
+-}
+
+    -- Transition matrix of an Markov Chain where each state is the
+    -- (idealized?) number of active slots in an EHCG window.
+    --
+    -- (This is a /stochastic matrix/ because every /row/ sums to 1. And every
+    -- state is reachable from every other, since @0 < asc@.)
```

Comment:
The second property is called "irreducibility", but it is not part of the definition of a stochastic matrix.

Rather, a Markov chain being irreducible and [aperiodic](https://en.wikipedia.org/wiki/Discrete-time_Markov_chain#Periodicity) (trivially satisfied, as actually we have non-zero probability of staying at a state for all states) is a sufficient criterion for there being a unique stationary distribution.

Reply:
Thanks, I added that vocab and stated that implication. 👍